### PR TITLE
DD4hep: Introduces functional Tracker navigation + DetIds building

### PR DESCRIPTION
**Pull request by ghugo83**
- There was no proper navigation when building Tracker geometry with DD4hep: non-matching volumes were treated in CmsTrackerLayerBuilder and CmsTrackerSubStrctBuilder, and the **Forward Tracker geometry builders were never reached**.
Hence, **no proper Tracker GeometricDet tree was actually created with DD4hep**.
- There was also no DetId construction. Once activated, there were **numerous regressions in DetIds**.

**This PR fixes these issues, which are absolute show-stoppers, located at the beginning of the sim step, for any workflow containing a Tracker.
This PR allows 1:1 results between old DD and DD4hep (for DetIds, geo placement, and sensor topology values).**
This PR is fully validated with the cmsExtended2021 scenario only.

NB: Ideally, I am not fully finished with this PR.
Since these issues were complete show-stoppers, I push it now, so that people can pull the commits.

### Issues solved
- _Problem:_ **Navigation, in Geometry/TrackerNumberingBuilder, does not work in DD4hep case.**
None of the attempts of using `while (fv.firstChild())` and `while ( fv.next(0))` in `CmsTrackerLevelBuilder::build` seem to work.
Furthermore, `fv.parent()` and` fv.nextSibling` have a different logic in the DD4hep case than old DD (because DD4hep itself has different logic), and hence cannot be used like with old DD (@ianna).

_Solution:_ **The key idea is to never use `while (fv.firstChild())` in `CmsTrackerLevelBuilder::build`, and instead take advantage of the calls stack.**
Each time` fv.firstChild() `would be evaluated, within a while loop in `CmsTrackerLevelBuilder::build`, the navigation would go to the next volume. **This would not be fine, because the CmsTrackerLevelBuilder::build calls are intricated (barrel::build calls layer::build, which in turn calls ladder::build, etc). 
Hence, the navigation would go too fast ahead, hence the present issues.**
Instead, **one should call for the next volume within CmsDetConstruction**. To go back up, make use of the **stack calls**.

- _Problem:_  **Once navigation is fixed, a segfault appears in CmsTrackerPhase1DiskBuilder.** 
(It was not appearing before, because navigation was not ok, hence the Forward Builder was never called.)

_Solution:_  Units are in cm with DD4hep, and there was a hardcoded value in 
https://github.com/cms-sw/cmssw/blob/master/Geometry/TrackerNumberingBuilder/plugins/CmsTrackerPhase1DiskBuilder.cc#L44 
`double radius_split = 100.;`
This resulted in one radius-split group being empty (values in cm were much smaller), hence the std::rotate call being not appropriate, hence the segfault.
-> **Keep consistent units in DetGeomDesc for all geo values + Protect the use of std::rotate.**

- _Problem:_  **DetIds in TEC (old DD versus DD4hep) were completely different.**

_Solution:_  **Remove non-desired custom navigation** in https://github.com/cms-sw/cmssw/blob/master/Geometry/TrackerNumberingBuilder/plugins/CmsTrackerRingBuilder.cc#L23 
**Just use a data member to access child name.** 
NB: Cannot be done within sortNS, as in there, the filtered view has already 'nexted'.

- _Problem:_  **DetIds in all Barrels (old DD versus DD4hep) were different.**

_Solution:_   Issue was at `float layerRadius = (det->params()[2] + det->params()[1]) / 2.;`
https://github.com/cms-sw/cmssw/blob/master/Geometry/TrackerNumberingBuilder/plugins/CmsTrackerLayerBuilder.cc#L53
**Order of shape parameters with DD4hep is different, leading to layerRadius not making sense in the DD4hep case.**
This, in turn, leads to sorting by layerRadius (hence DetIds) being inappropriate.
-> Keep same order of shape parameters for the few shapes of interest (box, trap, tubs).

- _Problem:_  **~half of sensors DetIds were wrong with a shift of 1.**

_Solution:_ The DD4hep filtered view is already filtered with `TkDDDStructure` in
https://github.com/cms-sw/cmssw/blob/master/Geometry/TrackerNumberingBuilder/plugins/DDDCmsTrackerContruction.cc#L57
Hence, `ExtractStringFromDDD<FilteredView>::getString(stereo, &fv) == "true"`, which filters through another attribute, was not properly evaluated.
https://github.com/cms-sw/cmssw/blob/master/Geometry/TrackerNumberingBuilder/plugins/CmsDetConstruction.cc#L16
Hence the shift in 1 in DetIds.
-> **Use GeometricDet getters instead. Of course, this is also true for Phase 2 sensors.**

### Remaining on the TO DO list, in view of an operational Tracker workflow
**- PRIORITY: Should allow namespace (+copy number?) in SpecPar sections** :
No namespace (and copy number?) is tolerated by DD4hep CMS FilteredView interface.
Hence, in the DD4hep case, the topology XML files need to be modified, to be able to be processed properly.
I think this is not desirable, and should be fixed in the DDCMS FilteredView.
_Motivations:_
- **Duplication of XML files** does not sound good.
- It is actually **not possible to have a working topology with no namespace for any Phase 2 Tracker**, because the module names (without namespace) are the same for OT and IT.
- This also leads, in the DD4hep case, to **null RadLength and Xi data members values**. More generally, **the paths in trackerRecoMaterial.xml are not properly recognized.**

**- Test all other geometries**
Once the point above is addressed, should compare the Tracker GeometricDets for ALL scenarios.
Only cmsExtended2021 comparison was done here.
Should also test the writer to DB.

**- Speed**
There seems to be a **major performance issue with DD4hep FilteredView::firstChild()** (?). This should be studied.
Geo is only constructed once, so not a dramatic issue. Though, there might hopefully be a way to improve this?
NB: This has nothing to do with the fixes in this PR. Performance issues were already present, and were not addressed here.

**- GeometricDetExtra**
In the DD4hepo case, there is no GeometricDetExtra handling. GeometricDetExtra  should be adapted or removed.

@ianna @vargasa @cvuosalo @civanch @dpiparo @mmusich

### CODE REVIEW

**Comment by ghugo83** [comment]
diff_2021
Diff of all GeometricDet data members (apart from RadLen and Xi), old DD versus DD4hep, 2021 scenario, now show a perfect match.

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
@ghugo83 - both namespace and copy numbers in `SpecPars` paths are allowed. Here is an example: 
https://cmssdt.cern.ch/lxr/source/SimTracker/TrackerMaterialAnalysis/plugins/dd4hep/DD4hep_ListGroups.cc#0381

**Comment by vargasa** [comment]
> **Failed tests: Build**

@smuzaffar, @mrodozov This is not related with the PR itself, the header was already included: https://cmssdt.cern.ch/lxr/source/DetectorDescription/DDCMS/interface/DDFilteredView.h#0023

**Comment by ghugo83** [comment]
Please wait, not done yet ;p

**Comment by ghugo83** [comment]
>@ghugo83 - both namespace and copy numbers in SpecPars paths are allowed. Here is an example:
https://cmssdt.cern.ch/lxr/source/SimTracker/TrackerMaterialAnalysis/plugins/dd4hep/DD4hep_ListGroups.cc#0381

@ianna Thanks for sharing this.
`std::vector<std::vector<cms::Node *>> children = fv1.children(k);`
Hmm but in https://github.com/cms-sw/cmssw/blob/master/SimTracker/TrackerMaterialAnalysis/plugins/dd4hep/DD4hep_ListGroups.cc#L381 , the paths are matched using no namespace, no?

Looking at https://github.com/cms-sw/cmssw/blob/master/DetectorDescription/DDCMS/src/DDFilteredView.cc#L303 
` std::vector<std::pair<std::string_view, int>> names = toNodeNames(selectPath);`
This relies on https://github.com/cms-sw/cmssw/blob/master/DetectorDescription/DDCMS/src/DDFilteredView.cc#L249 
std::vector<std::pair<std::string_view, int>> DDFilteredView::toNodeNames(const std::string& path)
which precisely seems to make the match with no namespace, no?
 `auto name = noNamespace(i);`

**Comment by ianna** [comment]
> > @ghugo83 - both namespace and copy numbers in SpecPars paths are allowed. Here is an example:
> > https://cmssdt.cern.ch/lxr/source/SimTracker/TrackerMaterialAnalysis/plugins/dd4hep/DD4hep_ListGroups.cc#0381
> 
> @ianna Thanks for sharing this.
> `std::vector<std::vector<cms::Node *>> children = fv1.children(k);`
> Hmm but in https://github.com/cms-sw/cmssw/blob/master/SimTracker/TrackerMaterialAnalysis/plugins/dd4hep/DD4hep_ListGroups.cc#L381 , the paths are matched using no namespace, no?
> 
> Looking at https://github.com/cms-sw/cmssw/blob/master/DetectorDescription/DDCMS/src/DDFilteredView.cc#L303
> ` std::vector<std::pair<std::string_view, int>> names = toNodeNames(selectPath);`
> This relies on https://github.com/cms-sw/cmssw/blob/master/DetectorDescription/DDCMS/src/DDFilteredView.cc#L249
> `std::vector<std::pair<std::string_view, int>> DDFilteredView::toNodeNames(const std::string& path)`
> which precisely seems to make the match with no namespace, no?
> `auto name = noNamespace(i);`
> https://github.com/cms-sw/cmssw/blob/master/DetectorDescription/DDCMS/src/DDFilteredView.cc#L312
> ` if (match(pathFromParent, names)) {`

@ghugo83 - thanks! It looks like it's unavoidable :-) I've opened an issue to remind ourselves https://github.com/cms-sw/cmssw/issues/31579

**Comment by ghugo83** [comment]
I wanted to do some restructuring, just did.

**Comment by ghugo83** [comment]
There was a build issue from RecoRomanPot/RecoFP420/ because it has an ill-formed BuildFile, but its dependency to Geometric was not needed anyway, hence it is removed.

**Comment by ghugo83** [comment]
In order to have operational SpecPars with DD4hep, trackerStructureTopology.xml for DD4hep is created, so that other people can now work on a workflow with Tracker with DD4hep (the 2021 one).
As mentioned in the PR description, we should definitely not go the way of duplicating XML files with SpecPars though, but only rely on the old DD ones, with no XML diff. Hence this file will be removed, once the CMS DD4hep FilteredView for DD4hep is improved.

**Comment by kpedro88** [comment]
@ghugo83 the Run 3 geometries are now managed similar to the Phase 2 geometries. Updates (to files such as `Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml`) must be made consistently using the relevant geometry script from https://github.com/cms-sw/cmssw/tree/master/Configuration/Geometry.

File: Geometry/TrackerNumberingBuilder/src/GeometricDet.cc [hunk 1]
```diff
@@ -124,54 +133,54 @@ GeometricDet::GeometricDet(const PGeometricDet::Item& onePGD, GeometricEnumType
...
+    params_.emplace_back(onePGD._params9);
+    params_.emplace_back(onePGD._params10);
   }
 
-  _ddd.reserve(onePGD._numnt);
-  _ddd.emplace_back(onePGD._nt0);
-  _ddd.emplace_back(onePGD._nt1);
-  _ddd.emplace_back(onePGD._nt2);
-  _ddd.emplace_back(onePGD._nt3);
+  ddd_.reserve(onePGD._numnt);
```
Comment (ianna) [inline Geometry/TrackerNumberingBuilder/src/GeometricDet.cc:165]: @ghugo83 - please, consider:
```c++
 auto nt = fv->navPos();
 ddd_ = nav_type(nt.begin(), nt.end());
```
https://cmssdt.cern.ch/lxr/source/DetectorDescription/DDCMS/interface/DDFilteredView.h#0206

File: Geometry/TrackerNumberingBuilder/src/GeometricDet.cc [hunk 2]
```diff
@@ -216,19 +225,19 @@ namespace {
...
-  std::for_each(_container.begin(), _container.end(), Deleter());
-  _container.clear();
+  std::for_each(container_.begin(), container_.end(), Deleter());
+  container_.clear();
 }
 
 GeometricDet::Position GeometricDet::positionBounds() const {
-  Position _pos(float(_trans.x() / cm), float(_trans.y() / cm), float(_trans.z() / cm));
-  return _pos;
+  Position pos_(float(trans_.x() / cm), float(trans_.y() / cm), float(trans_.z() / cm));
```
Comment (ianna) [inline Geometry/TrackerNumberingBuilder/src/GeometricDet.cc:233]: convertMmToCm?

**Comment by ianna** [comment]
> @ghugo83 the Run 3 geometries are now managed similar to the Phase 2 geometries. Updates (to files such as `Geometry/CMSCommonData/data/dd4hep/cmsExtendedGeometry2021.xml`) must be made consistently using the relevant geometry script from https://github.com/cms-sw/cmssw/tree/master/Configuration/Geometry.

@kpedro88 - it looks like there are no `dd4hep` Era scenarios defined in the dictionaries.

**Comment by civanch** [comment]
please test

let us test over night

**Comment by kpedro88** [comment]
@ianna the files in `Geometry/CMSCommonData/data/dd4hep` are generated by the geometry script. Therefore, they should only be modified via the geometry script.

(As I'm *sure* that I've explained before, the Eras in the geometry dictionaries are just recommendations or reminders for subdetector-specific modifiers. The actual workflows are defined separately. The dd4hep Era is not subdetector-specific, so it is not included in the dictionary recommendations. This does not impact the availability of a dd4hep workflow in any way. Though, I'm under the impression that a full dd4hep workflow is not functional yet.)

**Comment by ghugo83** [comment]
@kpedro88 The issue is that one XML file, used for DD4hep, is **different** than the XML file used for old DD. 
Hence it is _intentionally_ not listed in the dictionary, as, in my understanding, the files listed in the dictionary are **identical for old DD and DD4hep**.

This XML file is obviously temporary. It is done so that other people can use this branch and advance with a DD4hep Tracker.

The underlying actual issue, is **namespace treatment in SpecPar sections**. This is handled within the core DD4hep FilteredView CMSSW interface. **Basically, the match is presently made after calling a 'removeNamespace' function**.

There is nothing to do on the dictionary point of view for now, because I do not want to go to the path of having different XMLs for old DD and DD4hep anyway. 
Furthermore, for Phase 2 Tracker, having an extra XML will not even fully circumvent the namespace issue.

The plan is to **have a DD4hep FilteredView CMSSW interface, which can handle the existing XMLs, without any XML modification**.
I reported the issue on namespace treatment last week. 
Hence, my plan is that once a fix is incorporated to the FilteredView, I remove the extra XML.

File: Geometry/TrackerNumberingBuilder/src/GeometricDet.cc (hunk 1 above)
Comment (ghugo83) [reply Geometry/TrackerNumberingBuilder/src/GeometricDet.cc:165]: Hmm here, this is the constructor from PGeometricDet, not from the filtered view.
What you mention is done in the constructor from filtered view.

File: Geometry/TrackerNumberingBuilder/src/GeometricDet.cc (hunk 2 above)
Comment (ghugo83) [reply Geometry/TrackerNumberingBuilder/src/GeometricDet.cc:233]: yes  basically just changed all GeometricDet data members from _member to member_, so did not author this. But yes you are right, will use geant_units::operators::convertMmToCm

File: Geometry/TrackerNumberingBuilder/src/GeometricDet.cc (hunk 1 above)
Comment (ianna) [reply Geometry/TrackerNumberingBuilder/src/GeometricDet.cc:165]: @ghugo83 - you are right, I take it back :-)

**Comment by kpedro88** [comment]
@ghugo83 be that as it may, the next time somebody runs the 2021 geometry script, it will wipe out this difference. And the unit test will fail in the meantime (it doesn't fail in this PR because `Configuration/Geometry` has no explicit dependence on `Geometry/CMSCommonData/data/dd4hep`, which is a deficiency in the test setup.)

**Comment by ghugo83** [comment]
@kpedro88 yes as mentioned, this is temporary, I plan to update this file in this PR when is possible

File: Geometry/TrackerCommonData/data/PhaseI/dd4hep/trackerStructureTopology.xml [hunk 3]
```diff
@@ -0,0 +1,426 @@
+<?xml version="1.0"?>
```
Comment (ianna) [inline Geometry/TrackerCommonData/data/PhaseI/dd4hep/trackerStructureTopology.xml:1]: @ghugo83 - this file can be deprecated if you test this PR with https://github.com/cms-sw/cmssw/pull/31610

**Review by ianna — CHANGES_REQUESTED** [review]

**Comment by ianna** [comment]
@ghugo83 - BTW, the tests are incredibly verbose. It seems that the message logger configuration is not set correctly: all info from Algorithms is printed along with geometry and numbering  builders.

**Comment by ghugo83** [comment]
Yes will change the verbosity.
But was useful until now as I wanted to see what happens with the XML fix / debug performance of the DD4hep FilteredView.

**Comment by ianna** [comment]
@ghugo83 - nonzero `radLength` and `xi` :-)
```
GeometricDet TOBActiveRphi0
   RadLength 0.0289843
          Xi 6.91796e-05
GeometricDet TOBActiveSter0
   RadLength 0.0289843
          Xi 6.91796e-05

```

**Comment by ghugo83** [comment]
@ianna Hmm still not fully satisfied with the namespace fix unfortunately (for XML SpecPars).

For the pixel sensors paths it seems to work, though the SpecPars are not always properly recognized with DD4hep.

This leads to RadLen and Xi still null in TEC for example, when they should not:

**old DD:**
det->geographicalID() = 470095052
myDet->name() = TECModule5RphiActive
myDet->module->translation() = (-716.6416901,396.0738620,-1792.7510000)
**myDet->radLength() = 0.0457685
myDet->xi() = 0.0001035**

**DDhep**
myDet->geographicalID() = 470095052
myDet->name() = TECModule5RphiActive
myDet->module->translation() = (-716.6416901,396.0738620,-1792.7510000)
**myDet->radLength() = 0.0000000
myDet->xi() = 0.0000000**

**Comment by ianna** [comment]
Hmm... I get it from GeomDet:
```
GeometricDet tecmodule5:TECModule5RphiActive
   RadLength 0.07531
          Xi 0.000166242

```

**Comment by ghugo83** [comment]
@ianna Some sensors in TEC are fine, but not all. Look for DetId 470095052 ;)

**Comment by ghugo83** [comment]
Also
```
GeometricDet tecmodule5:TECModule5RphiActive
   RadLength 0.07531
          Xi 0.000166242
```
Has a different value than old DD.

RL_diff

As you see in the meld there is still a diff with Rl and Xi. So still some fun to have within the DD4hep FiltedredView (should not be much now though :))

**Comment by ghugo83** [comment]
@ianna The thing is that for the Tracker, tecmodule5:TECModule5RphiActive is not a unique identifier, so you need to use the DetId.

**Comment by ianna** [comment]
@ghugo83 and @cvuosalo - perhaps this PR could be merged ASIS? I think, it would help in debugging the SIM step1 issues in other sub detectors:
Screenshot 2020-10-01 at 09 13 04

**Comment by ghugo83** [comment]
Ok I will push the remaining commits today then.

**Comment by civanch** [comment]
@ghugo83 , you need proper rebase - something is wrong now

**Comment by ghugo83** [comment]
Basically the branch is this one:
https://github.com/cms-sw/cmssw/compare/master...ghugo83:DD4hep_Tracker_sim_temp_test?expand=1

**Comment by civanch** [comment]
@ghugo83 , it is not possible to change all categories

**Comment by ghugo83** [comment]
Hmm I do not understand why integrated commits from the release appear in the PR.
The diff of master with this exact branch is here: https://github.com/cms-sw/cmssw/compare/master...ghugo83:DD4hep_Tracker_sim_temp?expand=1 
I force-push that branch from local but it stays the same.

**Comment by ianna** [comment]
> Hmm I do not understand why integrated commits from the release appear in the PR.
> The diff of master with this exact branch is here: https://github.com/cms-sw/cmssw/compare/master...ghugo83:DD4hep_Tracker_sim_temp?expand=1
> I force-push that branch from local but it stays the same.

I think you need to rebase it

**Comment by ghugo83** [comment]
Ha it worked with a cms-init.
Sorry for that ;)
The issue was I had not done 
`git cms-init`
But directly pulled to my local, hence the latest X releases were not properly recognized.

**Comment by ianna** [comment]
please test

**Comment by ghugo83** [comment]
sorry for the branch issue

**Comment by ghugo83** [comment]
As advertised this PR fully fixes navigation and DetIds with DD4hep, for Phase 1 Tracker.

**Comment by ghugo83** [comment]
Ok so more seriously, on what remains in the TO DO list:

- As reported, to fix the RL/Xi parsing from XMLs +  many other parsing issues from XMLs in many other places (for example, notably, the parsing of Phase 2 Tracker XMLs), it is necessary to update the **DD4hep DDFilteredView.**
The DD4hep DDFilteredView is in the process of being fixed.
There are nice PRs on this topic here:
https://github.com/cms-sw/cmssw/pull/31621/files 
https://github.com/cms-sw/cmssw/pull/31614/files
They improve the situation, but presently the RL/Xi parsing from XMLs is still not ok, so there is still some additional effort to do here.

- Another point is the performance of the DD4hep filtered view.
It is not addressed here, as off-topic.
I reported several lagging points in DD4hep DDFilteredView in https://github.com/cms-sw/cmssw/issues/31579 
Basically, a huge part of the lagging seems to be also coming from `DDFilteredView::get<type>`.
By limiting the usage of regex + avoiding the use of `dd4hep::dd::split` there, there is already a significant improvement.
Still, it seems that there is additional big improvement possible with reasonable effort possible.

These points should not be put under the carper, they are important points to address, and believe also useful for other subdetectors / further steps.

**Comment by kpedro88** [comment]
+upgrade

**Comment by jfernan2** [comment]
+1

**Comment by ianna** [comment]
+1

**Comment by civanch** [comment]
+1

**Comment by ianna** [comment]
@ggovi and @slava77 - please, review.

File: CondTools/Geometry/plugins/PGeometricDetBuilder.cc [hunk 4]
```diff
@@ -189,7 +189,7 @@ void PGeometricDetBuilder::putOne(const GeometricDet* gd, PGeometricDet* pgd, in
     item._params9 = 0;
     item._params10 = 0;
   }
-  item._geographicalID = gd->geographicalID();
+  item._geographicalID = gd->geographicalId();
```
Comment (cvuosalo) [inline CondTools/Geometry/plugins/PGeometricDetBuilder.cc:192]: `geographicalID` is actually correct, but it is not worth changing back.

Comment (vargasa) [reply CondTools/Geometry/plugins/PGeometricDetBuilder.cc:192]: They serve the same purpose https://cmssdt.cern.ch/lxr/source/Geometry/TrackerNumberingBuilder/interface/GeometricDet.h#0151

File: Geometry/TrackerNumberingBuilder/interface/GeometricDet.h [hunk 5]
```diff
@@ -81,128 +80,118 @@ class GeometricDet {
...
-   */
-  void setGeographicalID(DetId id) { _geographicalID = id; }
+  // ACCESS GENERAL INFO
+  const std::string& name() const { return ddname_; }
+  const GeometricEnumType& type() const { return type_; }
+
+  // NAVIGATION related info
+  const nav_type& navType() const { return ddd_; }
+  NavRange navpos() const { return NavRange(&ddd_.front(), ddd_.size()); }
+  const DetId& geographicalId() const { return geographicalID_; }
```
Comment (vargasa) [inline Geometry/TrackerNumberingBuilder/interface/GeometricDet.h:96]: @cvuosalo is unified here

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/31573	93d3111 
- reco changes are rather technical, in a few files in RecoTracker/TkDetLayers `geographicalID` is changed to `geographicalId`, which seems like kind of random change https://github.com/cms-sw/cmssw/pull/31573#discussion_r499005238 
- jenkins tests pass and it looks like there are no meaningful/relevant changes
    - one example is https://tinyurl.com/y4svnzze in wf 23234.0

given still continuing review and the PR description containing "NB: Ideally, I am not fully finished with this PR.", this +1 may not be the last one, but we were asked to.

**Comment by ianna** [comment]
@ggovi - please, review.

File: Geometry/TrackerNumberingBuilder/interface/GeometricDet.h (hunk 5 above)
Comment (ghugo83) [reply Geometry/TrackerNumberingBuilder/interface/GeometricDet.h:96]: @cvuosalo Yes basically there were 2 member functions with different names, returning the same member data. I cleaned it on the fly.

**Comment by ggovi** [comment]
+1

**Comment by silviodonato** [comment]
+1
